lib/opam/opam_make_opam.ml

```diff
@@ -0,0 +1,162 @@
+open Core.Std
+open Core_extended.Std
+open Async.Std
+
+let format_cmds
+  : String.t List.t -> String.t =
+  fun cmds ->  "[\n" ^
+               (List.fold ~init:""
+                  ~f:(fun acc el -> acc ^ "    [" ^ el ^ "]\n") cmds) ^
+               "]\n"
+
+let create_build_cmds
+  : String.t List.t -> String.t =
+  function
+  | [] -> format_cmds ["make build"]
+  | build_cmds -> format_cmds build_cmds
+
+let create_install_cmds
+  : String.t List.t -> String.t =
+  function
+  | [] -> format_cmds ["make install"]
+  | install_cmds -> format_cmds install_cmds
+
+let create_remove_cmds
+  : String.t List.t -> String.t =
+  function
+  | [] -> format_cmds ["make remove"]
+  | remove_cmds -> format_cmds remove_cmds
+
+let format_build_dependencies
+  : String.t List.t -> String.t =
+  List.fold ~init:""
+    ~f:(fun acc el -> acc ^ " \"" ^ el ^ "\" {build}")
+
+let format_dependencies
+  : String.t List.t -> String.t =
+  List.fold ~init:""
+    ~f:(fun acc el -> acc ^ " \"" ^ el ^ "\"")
+
+let write_opam
+  : target_dir:String.t -> name:String.t -> semver:String.t
+  -> license:String.t -> maintainer:String.t Option.t -> author:String.t Option.t
+  -> homepage:String.t Option.t -> bug_reports:String.t Option.t -> dev_repo:String.t
+  -> build_cmds:String.t List.t -> install_cmds:String.t List.t
+  -> remove_cmds:String.t List.t -> depends:String.t List.t
+  -> build_depends:String.t List.t -> (Unit.t, Exn.t) Deferred.Result.t =
+  fun ~target_dir ~name ~semver ~license ~maintainer
+    ~author ~homepage ~bug_reports ~dev_repo ~build_cmds ~install_cmds
+    ~remove_cmds ~depends ~build_depends ->
+    let maintainer_str = match maintainer with
```

Comment:
You can abstract this apply-string-transformation-if-not-None patter into a function and get rid of all those matches